Added MacOS support

hsource opened the pull request:
## Motivation

I've been playing POE on Mac. Trading sucks as usual. I heard Awakened POE Trade was the best trade tool around, and it didn't seem too hard to port!

## Fixes

1. Upgrade uiohook-napi and electron-overlay-window for Mac support
   - electron-overlay-window update is pending [Added MacOS support #17](https://github.com/SnosMe/electron-overlay-window/pull/17). I'll update the package.json to refer to the new package once that's pulled/updated
2. Tweaked config a bit so that the default config works on Mac
   - Set `useOsGlobalShortcut: false` by default for Mac. For unknown reasons, the `globalShortcuts` hooks don't seem to ever get triggered on Mac
   - Have a smaller tray image
   - Disable code signing, mostly because I'm too lazy to set it up and it doesn't matter that much for Mac

## Testing

Tested with the game. It works pretty well!

https://user-images.githubusercontent.com/2937410/119623610-33c5c600-bdbd-11eb-92c7-f88a67acf900.mp4

SnosMe commented:
![image](https://user-images.githubusercontent.com/4292308/120197419-98ec4200-c229-11eb-9e3d-98f072aaa95d.png)

Noticed that overlay positioned incorrectly, this is issue in electron-overlay-window implementation for mac. It should report size of client area (without title bar)

hsource commented:
> ![image](https://user-images.githubusercontent.com/4292308/120197419-98ec4200-c229-11eb-9e3d-98f072aaa95d.png)
> 
> Noticed that overlay positioned incorrectly, this is issue in electron-overlay-window implementation for mac. It should report size of client area (without title bar)

I fixed this with the latest version of electron-overlay-window. I also changed this to stop using uiohook-napi, as the Mac version of libuiohook was pretty unreliable and kept giving me [this error](https://github.com/kwhat/jnativehook/issues/225). All that was needed to make globalShortcuts work was to remove references to CmdAndCtrl and just use Ctrl.

james-prado commented:
The price checker isn't working on the latest version of Big Sur and Monterey. The price checker or link to wiki commands aren't working no matter what keybind I use. I tried a bunch of different keybinds including ones that have proven to work for other commands in the app like shift + space, but none of them worked.

Samusia commented:
@james-prado, use the latest version (installed according to the instructions from the first message) on Monterey and work for me:
`control+d`:
![image](https://user-images.githubusercontent.com/8151745/139601506-e6b286df-5550-40b8-9967-130bec256856.png)
As well as `option+w` and "map check"

james-prado commented:
@Samusia did these tests on the latest release (v2.14) following all the installation steps. I also checked the logs and every time i try to interact with an item I get the following error `warn [clipboard]: No item text found. {"text":""}` so I guess it has something to do with the accessing the clipboad?

james-prado commented:
Seems I also cant build or serve locally on any m1 macs:
```
 DONE  Compiled successfully in 1156ms                                                                                                                                                      6:34:49 p.m.

  File                      Size                                                                                 Gzipped

  dist_electron/index.js    1374.52 KiB                                                                          281.44 KiB

  Images and other types of assets omitted.

 INFO  Launching Electron...
App threw an error during load
Error: No native build was found for platform=darwin arch=arm64 runtime=electron abi=89 uv=1 armv=8 libc=glibc node=14.17.0 electron=14.0.1
    loaded from: /Users/jamesprado/Git/awakened-poe-trade/node_modules/uiohook-napi
    at Function.load.path (/Users/jamesprado/Git/awakened-poe-trade/node_modules/node-gyp-build/index.js:59:9)
    at load (/Users/jamesprado/Git/awakened-poe-trade/node_modules/node-gyp-build/index.js:21:30)
    at Object.<anonymous> (/Users/jamesprado/Git/awakened-poe-trade/node_modules/uiohook-napi/dist/index.js:6:38)
    at Module._compile (internal/modules/cjs/loader.js:1083:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1113:10)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Module._load (internal/modules/cjs/loader.js:781:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12913)
    at Module.require (internal/modules/cjs/loader.js:964:19)
    at require (internal/modules/cjs/helpers.js:88:18)
```

james-prado commented:
> @Samusia did these tests on the latest release (v2.14) following all the installation steps. I also checked the logs and every time i try to interact with an item I get the following error `warn [clipboard]: No item text found. {"text":""}` so I guess it has something to do with the accessing the clipboad? The app has accessibility permissions though, so idk why it wouldn't be able to access my clipboard

Turns out it was an issue with another app stealing the ctrl + alt + c keybinding which is needed for awakened poe.

Samusia commented:
@hsource, is there a chance to fix next error:
```
Uncaught Exception:
Error: Invalid key flag specified.
at B$ (/Applications/Awakened PoE Trade.app/Contents/Resources/app.asar/main.js:111:11646)
at Function.<anonymous> (/Applications/Awakened PoE Trade.app/Contents/Resources/app.asar/main.js:111:10897)
```
![image](https://user-images.githubusercontent.com/8151745/169387844-76ddb317-1001-409c-a9ba-252430cdf407.png)

AndersonDev commented:
Hi! I beg you to fix the error pointed higher

Samusia commented:
@hsource I tried the latest version: [v3.17.10005.2](https://github.com/hsource/awakened-poe-trade/releases/download/v3.17.10005.2/Awakened.PoE.Trade-3.17.10005.2.dmg), but the error still remained.
<img width="244" alt="image" src="https://user-images.githubusercontent.com/8151745/169642355-b768dc95-050f-48f4-a94b-50a683fcd12c.png">

Samusia commented:
@hsource Thank you sir!!!
The price check works as it should.
If you have time, can you also check why `F5` does not work? It's not as relevant as checking the price, but it would be great if that worked too.
<img width="246" alt="image" src="https://user-images.githubusercontent.com/8151745/169644121-f3d9f59d-483b-490f-9e66-03d7456fff9c.png">

AndersonDev commented:
@hsource Hi! May I ask you to get a look at why prefixes/suffixes are not passed? I'll be very thankful!

Samusia commented:
@SnosMe is there any chance that this PR will be added to the release?

SnosMe commented:
If there was no chance, I would close it already. 😄

AndersonDev commented:
@SnosMe We don't have anyone interested with macOS / virtual machine to make it work? I'm not super familiar with this stack, but currently can play only on macOS. Without your brilliant solution it's extremely hard.

greenhost87 commented:
Any news on this?

phen0menon commented:
I forked the repository, created a fix and released the `.dmg` file for installation. It works for me and my friends PERFECTLY and possibly for others who replied under the thread: https://www.pathofexile.com/forum/view-thread/3333212/page/1

Check the thread and try step-by-step guide to install Awakened for MacOS.

message commented:
@phen0menon did you push your PR changes?

greenhost87 commented:
@phen0menon Do you have any plans to update this PR? or update you repo in case of 3.21 launch?

phen0menon commented:
@greenhost87 Yep, I'm gonna update once 3.21 released. Kind of busy now, sorry ;(

AndersonDev commented:
Why it can't be tested and merged in dev?

AndersonDev commented:
@phen0menon Can we kindly ask you to update it?

greenhost87 commented:
Any progress with merge that?

SnosMe commented on main/src/shortcuts/Shortcuts.ts:
```diff
@@ -223,15 +223,27 @@ export class Shortcuts {
+    UiohookKey.C,
+    // On Mac, robotjs requires the modifiers to be specified in this way to
+    // register. See https://github.com/octalmage/robotjs/issues/208#issuecomment-223828356
+    isMac ? modifierKeys : undefined
```
I don't use robotjs anymore, are u sure this is required? Becuase it will mean you can't chain price checks like
Ctrl + D, D, D, D, D

hsource replied:
Actually, I tested a bit more. It looks like we don't actually need to add the modifier keys on the `keyTap` call. I did discover that on Mac, the `pressedModKeys` still need to be toggled to copy Advanced Mod Descriptions for some reason. I don't totally understand why, but it works!

hsource commented:
I just re-tested the branch. I think it's pretty well-functioning and good to merge!

Here are things I confirmed to work (keyboard shortcuts)

- Settings and saving
- Price check auto-hide
- Price check no auto-hide
- Map check/item info
- Client log and game config default directories
- Adding new searches
- Timer

Here are the things that are broken and why they're a bit harder to fix:

- Adding new image strip. When we try to select an image, it opens a new window that's technically not owned by PoE or the overlay, which causes the overlay to disable since it thinks it's no longer in PoE. I need to change `electron-overlay-window` to add an exception for the File selection dialog
- App doesn't exit when using the menubar. I'm not exactly sure why it's getting stuck, so I'll need to debug a bit more

I think it's good enough to merge, so if you're OK with it, I'd appreciate it!

itslooklike commented:
Thanks for adding macOS support guys! you are very cool!

dinfyru commented:
![image](https://github.com/SnosMe/awakened-poe-trade/assets/10163112/8a637262-30d2-42a2-aab9-28a808b5a2c6)
I have error when running it on MacBook M1 Max
[Poe Trade log.txt](https://github.com/SnosMe/awakened-poe-trade/files/13797428/Poe.Trade.log.txt)

dinfyru commented:
Don't crash if open with Rosetta. But always ask this and don't run
![image](https://github.com/SnosMe/awakened-poe-trade/assets/10163112/563d1bc7-d2d5-4eaa-b2a2-9c81d2cafecd)

SaloEater commented:
> Don't crash if open with Rosetta. But always ask this and don't run ![image](https://private-user-images.githubusercontent.com/10163112/293451593-563d1bc7-d2d5-4eaa-b2a2-9c81d2cafecd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNTY0NDcsIm5iZiI6MTcxMjE1NjE0NywicGF0aCI6Ii8xMDE2MzExMi8yOTM0NTE1OTMtNTYzZDFiYzctZDJkNS00ZWFhLWIyYTItOWM4MWQyY2FmZWNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAzVDE0NTU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4MWU3MTY0N2UxY2QyNDU2NzZkNjk2ZGYwNzAxMzY0ZjVlZDhiNDVmZjAyMjM0YjNmNzNhNWUwMGIwZTdlZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.r6WZih-_Yt9Z2KG-w1401CnMRbfwNK9V3OoAertQ5kU) ![image](https://private-user-images.githubusercontent.com/10163112/293451748-923f6312-36b5-4ef0-a3e9-fc7d0ae1a1c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNTY0NDcsIm5iZiI6MTcxMjE1NjE0NywicGF0aCI6Ii8xMDE2MzExMi8yOTM0NTE3NDgtOTIzZjYzMTItMzZiNS00ZWYwLWEzZTktZmM3ZDBhZTFhMWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAzVDE0NTU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhNDgyNWQ4NDljMzg0OGQ0NmEyODIzMDQ4ZTc2YWRhMTYzM2NhMTljZjUxYmJkMjU3ZTEzMjAwZTg4MWMyZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.R4PHkJ0nQW25yWoMPqXHa2ooVcwh0eh0oEjbq7s676o)

Hey, did you find any solution to this?
